Aleo schnorr

anstylian opened the pull request.

anstylian committed:
support aleo schnorr

anstylian committed:
use aleo structures to sign

anstylian committed:
sign group

anstylian committed:
unit test

anstylian committed:
cleanning

tilacog commented on src/aleo_schnorr/mod.rs:
```diff
@@ -0,0 +1,138 @@
+        assert!(success);
+    }
+
+    /// Check keygen/signing outputs against golden files to catch regressions (such as on updating deps).
+    /// Golden files were generated from tofn commit corresponding to tofnd v0.10.1 release
+    #[test]
+    fn keygen_sign_known_vectors() {
+        // todo!()
+    }
```
Do we plan on implementing this?

If not, I'd suggest we remove this test item entirely. 

If we are planning to implement it, I suggest using the [ignored attribute](https://doc.rust-lang.org/reference/attributes/testing.html#the-ignore-attribute). This approach is more transparent since the test's ignored status will be visible during test runs, which is better than having a silent no-op test.

```suggestion
    /// Check keygen/signing outputs against golden files to catch regressions (such as on updating deps).
    /// Golden files were generated from tofn commit corresponding to tofnd v0.10.1 release
    #[test]
    #[ignore = "not yet implemented"]
    fn keygen_sign_known_vectors() { }
```

tilacog commented on src/aleo_schnorr/mod.rs:
```diff
@@ -0,0 +1,138 @@
+    signing_key: &KeyPair<N>,
+    msg_to_sign: &MessageDigest,
+) -> TofnResult<BytesVec> {
+    let message = aleo_encoded(msg_to_sign).map_err(|_| TofnFatal)?;
```
Quick question: 
Do we need this `.map_err` call? It seems this fn and `aleo_encoded` share the error type, so I wonder if just using a single `?` operator would be sufficient.

tilacog reviewed (approved):
LGTM. 
I just left a few minor comments.

anstylian committed:
apply review comments

vaporif commented on src/aleo_schnorr/mod.rs:
```diff
@@ -0,0 +1,137 @@
+        session_nonce,
+    )?;
+
+    let private_key = PrivateKey::new(&mut rng).map_err(|_| TofnFatal)?;
```
how about just change to 
```
.map_err(|_| { 
  error!("failure to to generate private key");
  TofnFatal
})?;
```

like in `ecdsa` for consistency
https://github.com/eigerco/tofn/blob/dfaab09f9b9cc0bee52c4ddb15f0e893fd6fec27/src/ecdsa/mod.rs#L53-L56

vaporif commented on src/aleo_schnorr/mod.rs:
```diff
@@ -0,0 +1,137 @@
+    )?;
+
+    let private_key = PrivateKey::new(&mut rng).map_err(|_| TofnFatal)?;
+    let aleo_account = Account::try_from(private_key).map_err(|_| TofnFatal)?;
```
`error!` for consistency
https://github.com/eigerco/tofn/blob/dfaab09f9b9cc0bee52c4ddb15f0e893fd6fec27/src/ecdsa/mod.rs#L53-L56

vaporif commented on src/aleo_schnorr/mod.rs:
```diff
@@ -0,0 +1,137 @@
+    let sign = signing_key
+        .aleo_account
+        .sign(&message, &mut rand_chacha::ChaChaRng::from_entropy())
+        .map_err(|_| TofnFatal)?;
```
`error!` for consistency
https://github.com/eigerco/tofn/blob/dfaab09f9b9cc0bee52c4ddb15f0e893fd6fec27/src/ecdsa/mod.rs#L84-L86

vaporif commented on src/aleo_schnorr/mod.rs:
```diff
@@ -0,0 +1,137 @@
+    signing_key: &KeyPair<N>,
+    msg_to_sign: &MessageDigest,
+) -> TofnResult<BytesVec> {
+    let message = aleo_encoded(msg_to_sign).map_err(|_| TofnFatal)?;
```
`error!` for consistency
https://github.com/eigerco/tofn/blob/dfaab09f9b9cc0bee52c4ddb15f0e893fd6fec27/src/ecdsa/mod.rs#L84-L86

vaporif commented on src/aleo_schnorr/mod.rs:
```diff
@@ -0,0 +1,137 @@
+
+    let message = aleo_encoded(message)?;
+
+    let address = Address::from_str(address).map_err(|_| TofnFatal)?;
```
`error!` for consistency
https://github.com/eigerco/tofn/blob/dfaab09f9b9cc0bee52c4ddb15f0e893fd6fec27/src/ecdsa/mod.rs#L84-L86

vaporif commented on src/aleo_schnorr/mod.rs:
```diff
@@ -0,0 +1,137 @@
+}
+
+fn aleo_encoded<N: Network>(data: &MessageDigest) -> TofnResult<Vec<Field<N>>> {
+    let data = data.as_ref().try_into().map_err(|_| TofnFatal)?;
```
`error!` for consistency
https://github.com/eigerco/tofn/blob/dfaab09f9b9cc0bee52c4ddb15f0e893fd6fec27/src/ecdsa/mod.rs#L84-L86

vaporif commented on src/aleo_schnorr/mod.rs:
```diff
@@ -0,0 +1,137 @@
+    let message = format!("{num}group");
+
+    snarkvm::prelude::Value::from_str(message.as_str())
+        .map_err(|_| TofnFatal)?
```
`error!` for consistency
https://github.com/eigerco/tofn/blob/dfaab09f9b9cc0bee52c4ddb15f0e893fd6fec27/src/ecdsa/mod.rs#L84-L86

vaporif commented on src/aleo_schnorr/mod.rs:
```diff
@@ -0,0 +1,137 @@
+    snarkvm::prelude::Value::from_str(message.as_str())
+        .map_err(|_| TofnFatal)?
+        .to_fields()
+        .map_err(|_| TofnFatal)
```
`error!` for consistency
https://github.com/eigerco/tofn/blob/dfaab09f9b9cc0bee52c4ddb15f0e893fd6fec27/src/ecdsa/mod.rs#L84-L86

anstylian committed:
enhance error messages

anstylian committed:
use tracing::error;

anstylian committed:
update snarkvm dependencies

anstylian committed:
feat(aleo-schnorr): refactor API, add test vectors, enable by default

- Refactored Aleo Schnorr key and signature API to use byte encoding for keys and signatures.
- Added `sign_with_rng` for deterministic testing.
- Changed `verify` to accept byte-encoded addresses and signatures.
- Added golden test vectors for keygen/signing to catch regressions.
- Enabled `aleo-schnorr` feature by default in Cargo.toml.

anstylian committed:
feat(aleo-schnorr): refactor switch from sign/verify bytes, to sign/verify snarkvm group

anstylian committed:
feat(aleo-schnorr): update to snarkvm 4.1.0 to avoid rocksdb dependency

Update Aleo Schnorr dependencies to use snarkvm-console 4.1.0 and snarkos-account 4.1.0, removing the need for the snarkvm crate and its transitive dependency on rocksdb. Bring only what is needed from snarkvm.

anstylian committed:
Merge pull request #3 from eigerco/refactor-aleo-schnorr

feat(aleo-schnorr): refactor API, add test vectors, enable by default



- Refactored Aleo Schnorr key and signature API to use byte encoding for keys and signatures.
- Added sign_with_rng for deterministic testing.
- Changed verify to accept byte-encoded addresses and signatures.
- Added golden test vectors for keygen/signing to catch regressions.
- Enabled aleo-schnorr feature by default in Cargo.toml.

anstylian committed:
feat(aleo-schnorr): refactor for Leo compatibility & improve signing API (#4)

* feat(aleo-schnorr): refactor for Leo compatibility & improve signing API

- Refactor Aleo Schnorr signature module for direct compatibility with Leo programs.
- Add detailed module docs explaining the bytes→Group→Value→Fields→Signature transformation.
- Introduce msg_to_fields helper for consistent message conversion.
- Update test vectors for new signature outputs.
- Minor: fix conditional compilation in rng.rs for tests.
- use snarkos-account 4.1.0 (snarkvm-with-cosmwasm tag)
